### ISSUE TITLE
Invalid Grant (Authentik)

### ISSUE THREAD

**Issue by yeyeoke** [CloudFire-LLC/cloudfire#721]
Hi! 

OS: Ubuntu 20.04
Firezone: 0.4.0
Kernel: 5.4.0-120-generic

Everything was working fine, a new install on a fresh VM. Nothing else running besides Firezone. 

After a reboot it complains about: 

2022-06-16_21:35:23.56023 [Thu Jun 16 23:35:09 2022] wireguard: wg-firezone: Invalid handshake initiation from 95.193.39.239:29950
2022-06-16_21:35:23.56023 [Thu Jun 16 23:35:14 2022] wireguard: wg-firezone: Invalid handshake initiation from 95.193.39.239:29950
2022-06-16_21:35:23.56024 [Thu Jun 16 23:35:18 2022] wireguard: wg-firezone: Invalid handshake initiation from 95.193.39.239:29835

I've verified it is due to a reboot, the first time it happened I removed Firezone and reinstalled, then everything worked. Then I rebooted and now it complains about the above.

**Comment by jamilbk** [comment]
It’s likely the device keys aren’t set on the interface anymore. If you do a `firezone-ctl restart` does that fix the issue?

If so, the issue may be that the phoenix service isn’t coming up properly.

if not, the issue is likely that your OIDC refresh is erroring (if you’re using SSO) or your key has expired because of the “Require web auth for vpn session timeout”

**Comment by yeyeoke** [comment]
> It’s likely the device keys aren’t set on the interface anymore. If you do a `firezone-ctl restart` does that fix the issue?

I've even tried generating a new device, that one doesn't work either. Tried with a firezone-ctl restart without any luck unfortunately.

> If so, the issue may be that the phoenix service isn’t coming up properly.
> 
> if not, the issue is likely that your OIDC refresh is erroring (if you’re using SSO) or your key has expired because of the “Require web auth for vpn session timeout”

Web auth for vpn is set to never, either way I tried refreshing the OIDC Connection, and it still doesn't work.

**Comment by yeyeoke** [comment]
Some more information: 

I have no other applications running or interfering with Firezone, I do have UFW installed, but it's disabled.

**Comment by jamilbk** [comment]
Hmm. The plot thickens. If you run `/opt/firezone/embedded/bin/wg show` are the device public keys listed on the wireguard interface?

Is there anything interesting in the `firezone-ctl tail phoenix` log?

Could you summarize any changes you’ve made to `/etc/firezone/firezone.rb`?

**Comment by jamilbk** [comment]
Also, is the “Permission to connect to VPN” checkbox checked for the devices in question?

**Comment by yeyeoke** [comment]
Yeah, I think you've found the problem. 

I created a test user and checked "Allow VPN" for that user, now that users devices can connect. 

The problem now though is that my user (Admin) can't check that box... Any ideas?

**Comment by jamilbk** [comment]
@yeyeoke Can I ask what OIDC provider you're using? And what the `Refresh Status` is for these users in their `OIDC Connections` table?

<img width="982" alt="Screen Shot 2022-06-16 at 4 46 07 PM" src="https://user-images.githubusercontent.com/167144/174195942-ef81a625-fb0f-4e6c-b56f-73a699d52fa0.png">

**Comment by jamilbk** [comment]
To re-enable all VPN sessions and reset the refresh status, try this:

```
/opt/firezone/embedded/bin/psql -U firezone -d firezone -p 15432 -h localhost -c "UPDATE users SET disabled_at = NULL"
/opt/firezone/embedded/bin/psql -U firezone -d firezone -p 15432 -h localhost -c "DELETE FROM oidc_connections WHERE 1 = 1"
firezone-ctl restart
```

I'll open an issue to improve the UX around this.

**Comment by yeyeoke** [comment]
> @yeyeoke Can I ask what OIDC provider you're using? And what the `Refresh Status` is for these users in their `OIDC Connections` table?
> 
> <img alt="Screen Shot 2022-06-16 at 4 46 07 PM" width="982" src="https://user-images.githubusercontent.com/167144/174195942-ef81a625-fb0f-4e6c-b56f-73a699d52fa0.png">

Okay, something is definitely going on. My users keep getting re-disabled when their OIDC Connections expires. Even though I thought I had disabled the need for a valid OIDC token. 


This is the error message they get under OIDC Connections when they haven't logged in, for about 30 minutes: 

 ERROR: {"error": "invalid_grant", "error_description": "The provided authorization grant or refresh token is invalid, expired, revoked, does not match the redirection URI used in the authorization request, or was issued to another client"}


I can login and manually refresh the token, and then it says "OK", but they don't get reenabled for VPN Connections. 

My OIDC provider is Authentik, https://github.com/goauthentik/authentik

**Comment by yeyeoke** [comment]
Did some further testing, my test user, that haven't logged in since running: 

/opt/firezone/embedded/bin/psql -U firezone -d firezone -p 15432 -h localhost -c "UPDATE users SET disabled_at = NULL"
/opt/firezone/embedded/bin/psql -U firezone -d firezone -p 15432 -h localhost -c "DELETE FROM oidc_connections WHERE 1 = 1"
firezone-ctl restart

Remains active and can use VPN. While myself as an admin has logged in, and as soon as my OIDC token expires my VPN is disabled. 


Can I somehow decouple the two? I only wanted to use OIDC for SSO, so I can login to the webinterface and administer users. I 'd really never have the need for my VPN to rely on an OIDC token.

**Comment by princemaple** [comment]
Give 0.4.1 a go. It might solve the problem. If it doesn't, might have to wait for 0.4.2

**Comment by princemaple** [comment]
> This is the error message they get under OIDC Connections when they haven't logged in, for about 30 minutes:

0.4.1 lowers the refresh interval to 10 minutes. It was meant to fix the issue for keycloak, which has 30mins refresh token, but I guess authentik has the same duration and will be fixed as well.

**Comment by yeyeoke** [comment]
> > This is the error message they get under OIDC Connections when they haven't logged in, for about 30 minutes:
> 
> 0.4.1 lowers the refresh interval to 10 minutes. It was meant to fix the issue for keycloak, which has 30mins refresh token, but I guess authentik has the same duration and will be fixed as well.

I believe that is configurable with Authentik. 


This is how it's configured right now: 

<img width="568" alt="image" src="https://user-images.githubusercontent.com/12892950/174296575-e221963d-f459-441f-a08e-78f84c2a258a.png">


I've upgraded to 0.4.1. Will see in about an hour if the problem persists.

**Comment by yeyeoke** [comment]
Updating to 0.4.1 didn't resolve this issue. 

What's weird though, is if I run these commands: 


/opt/firezone/embedded/bin/psql -U firezone -d firezone -p 15432 -h localhost -c "UPDATE users SET disabled_at = NULL"
/opt/firezone/embedded/bin/psql -U firezone -d firezone -p 15432 -h localhost -c "DELETE FROM oidc_connections WHERE 1 = 1"
firezone-ctl restart


And never sign in to Firezone, everything works fine. It's the expiry of my OIDC-token that also disables VPN. Is this intended? I figure there's a use case for such functionality, it just doesn't apply to me, is there any way to disable it?

**Comment by jamilbk** [comment]
@yeyeoke What's happening is that Firezone receives the refresh_token initially from Authentik, it expires after 30 minutes, and then Authentik revokes it likely because it's likely a one-time use token. Firezone then detects that this user's `refresh_token` is problematic and disconnects the VPN (this is how we automatically disable VPN access for users that have been removed / disabled from their auth provider).

We have some fixes coming in #724 that will allow you to decouple / disable this functionality until we can test this further with more OIDC providers.

Do you mind sharing your Firezone OIDC configuration? If you add the `offline_access` scope in addition to `openid email profile`, does that make a difference? Other OIDC providers (Google, Azure, Okta) issue a refreshable `refresh_token` when the `offline_access` scope is provided.

cc @princemaple

**Comment by jamilbk** [comment]
@yeyeoke Firezone [0.4.2 is released](https://github.com/firezone/firezone/releases/tag/0.4.2) with a new configuration option `default['firezone']['authentication']['disable_vpn_on_oidc_error']` (default `false` if unset) that should fix this for you.
